lib/bolt/application.rb

```diff
@@ -0,0 +1,613 @@
+    # @param targets [Array[String], NilClass] The targets to pass to the plan.
+    # @param params [Hash] Parameters to pass to the plan.
+    # @return [Bolt::PlanResult]
+    #
+    def plan_run(plan, targets, params: {})
+      if targets
+        if params['nodes'] || params['targets']
+          key = params.include?('nodes') ? 'nodes' : 'targets'
+          raise Bolt::CLIError,
+                "A plan's '#{key}' parameter can be specified using the --#{key} option, but in that " \
+                "case it must not be specified as a separate #{key}=<value> parameter nor included " \
+                "in the JSON data passed in the --params option"
+        end
+
+        plan_params  = pal.get_plan_info(plan)['parameters']
+        target_param = plan_params.dig('targets', 'type') =~ /TargetSpec/
+        node_param   = plan_params.include?('nodes')
+
+        if node_param && target_param
+          msg = "Plan parameters include both 'nodes' and 'targets' with type 'TargetSpec', " \
+                "neither will populated with the value for --nodes or --targets."
+          Bolt::Logger.warn("nodes_targets_parameters", msg)
+        elsif node_param
+          params['nodes'] = targets.join(',')
+        elsif target_param
+          params['targets'] = targets.join(',')
+        end
+
+        inventory.get_targets(targets)
+      end
+
+      plan_context = { plan_name: plan, params: params }
+
+      executor.start_plan(plan_context)
+      result = pal.run_plan(plan, params, executor, inventory, plugins.puppetdb_client)
+      executor.finish_plan(result)
+
+      result
+    end
+
+    # TODO: This should probably be separate #show_plan and #list_plans methods.
+    #
+    # Show plan information.
+    #
+    # @param plan [String, NilClass] The name of the plan to show.
+    # @param filter [String, NilClass] A string to filter plans by.
+    # @return [Hash]
+    #
+    def plan_show(plan, filter: nil)
+      if plan
+        pal.get_plan_info(plan)
+      else
+        plans = filter_content(pal.list_plans_with_cache(filter_content: true), filter)
+        { plans: plans, modulepath: pal.user_modulepath }
+      end
+    end
+
+    # Show available plugins.
+    #
+    # @return [Hash]
+    #
+    def plugin_show
+      { plugins: plugins.list_plugins, modulepath: pal.user_modulepath }
+    end
+
+    # Initialize the current directory as a Bolt project.
+    #
+    # @param name [String] The name of the project.
+    # @param [Bolt::Outputter] An outputter instance.
+    # @param modules [Array[String], NilClass] Modules to install.
+    # @return [Boolean]
+    #
+    def project_init(name, outputter, modules: nil)
+      Bolt::ProjectManager.new(config, outputter, pal)
+                          .create(Dir.pwd, name, modules)
+    end
+
+    # Migrate a project to current best practices.
+    #
+    # @param [Bolt::Outputter] An outputter instance.
+    # @return [Boolean]
+    #
+    def project_migrate(outputter)
```

Comment:
Not really sure what this is

Reply:
The method itself? It implements `bolt project migrate`, which is a command that will update your Bolt project to be in line with what the current version of Bolt is expecting. More info is here: https://puppet.com/docs/bolt/latest/projects.html#migrate-a-bolt-project